modules/hitl2/src/backend/middleware.ts

```diff
@@ -0,0 +1,127 @@
+import * as sdk from 'botpress/sdk'
+import _ from 'lodash'
+import LRU from 'lru-cache'
+
+import { EscalationType } from './../types'
+import { StateType } from './index'
+import Repository from './repository'
+import Socket from './socket'
+
+const registerMiddleware = async (bp: typeof sdk, state: StateType) => {
+  const realtime = Socket(bp)
+  const repository = new Repository(bp)
+  const cache = new LRU<string, string>({ max: 1000, maxAge: 1000 * 60 * 60 * 24 }) // 1 day
+
+  const pipeEvent = async (event: sdk.IO.IncomingEvent, target: string, threadId: string) => {
+    bp.events.sendEvent(
+      bp.IO.Event({
+        botId: event.botId,
+        target: target,
+        threadId: threadId,
+        channel: event.channel,
+        direction: 'outgoing',
+        type: event.type,
+        payload: event.payload
+      })
+    )
+  }
+
+  const getEscalation = (cache, botId, threadId) => {
+    return cache.get(_.join([botId, threadId], '.'))
+  }
+
+  const cacheEscalation = async (botId: string, threadId: string, escalation: EscalationType) => {
+    cache.set(_.join([botId, threadId], '.'), escalation.id)
+  }
+
+  const expireEscalation = async (botId: string, threadId: string) => {
+    cache.del(_.join([botId, threadId], '.'))
+  }
+
+  const incomingHandler = async (event: sdk.IO.IncomingEvent, next: sdk.IO.MiddlewareNextCallback) => {
+    if (event.type !== 'text') {
+      return next()
+    }
+
+
+    const escalationId = getEscalation(cache, event.botId, event.threadId)
+
+    // Only escalations with status 'pending' or 'assigned' are cached
+    if (!escalationId) {
+      next(undefined, false)
+      return
+    }
+
+    const escalation = await repository
+      .escalationsQuery(builder => {
+        builder.where('id', escalationId)
+      })
+      .then(data => _.head(data))
```

Comment:
Can  you add helper methods directly in the repository instead? Ex: `const escalation = await repository.getById(escalationId')`. This way we keep all the SQL logic in the same file, and the usage is simpler